---
title: "Diabroticum Insect Differential expression"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Classify transcripts into insect, plant and bacteria

```{r classification}

#list.of.packages <- c("here", "VennDiagram","ape","gplots")
#new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
#if(length(new.packages)) install.packages(new.packages, repos = "http://cran.us.r-project.org")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager",repos = "http://cran.us.r-project.org")
library(here)

dat=read.table(here("Final_classification.txt"),header=T,sep='\t',stringsAsFactors = F)

###Load Genus to higher end classification
dat1=unique(dat)
dat1$key=dat1$genus
dat1$key=ifelse(is.na(dat1$key)|dat1$key=="",dat1$species,dat1$key)

dat1$value=dat1$phylum
dat1$value=ifelse(dat1$superkingdom=="Bacteria","Bacteria",dat1$value)
dat1$value=ifelse(dat1$superkingdom=="Archaea","Archaea",dat1$value)
dat1$value=ifelse(dat1$kingdom=="Viridiplantae","Viridiplantae",dat1$value)
dat1$value=ifelse(dat1$kingdom=="Fungi","Fungi",dat1$value)
dat1$value=ifelse(dat1$phylum=="Ascomycota"|dat1$phylum=="Basidiomycota","Fungi",dat1$value)
dat1$value=ifelse(grepl("virus",dat1$key),"Virus",dat1$value)
dat2=dat1[,6:7]
dat3=unique(dat2)
dat4=dat3[dat3$value%in%c("Nematoda","Archaea","Arthropoda","Viridiplantae","Chordata","Fungi"),]
dat4=dat4[dat4$key!="",]

### load modified transcript annotations & process only swissprot evidence
dt=read.table(here("annotations_extracted2.txt"),header=T,sep='\t',stringsAsFactors = F)

dt.x=dt[,c(1,4)]
dt1=dt.x[dt.x$Swissprot.Description!=" NA",]

bact=dt1$Genes[grepl("Bacteria", dt1$Swissprot.Description,ignore.case = TRUE)]
plant=dt1$Genes[grepl("Viridiplantae", dt1$Swissprot.Description,ignore.case = TRUE)]
arthropoda=dt1$Genes[grepl("Arthropoda", dt1$Swissprot.Description,ignore.case = TRUE)]

```

## Including Plots Venndiagram


```{r venn, echo=FALSE}
system("rm *.log")
system("rm *.tiff")
library(VennDiagram)
all=unique(dt$Genes)
ann=unique(c(bact,plant,arthropoda))
unk=length(all)-length(ann)
unk
arth=gsub(" ", "", arthropoda)
bac=gsub(" ", "", bact)
pla=gsub(" ", "", plant)

lst=list()
lst[1]=list(bac)
lst[2]=list(pla)
lst[3]=list(arth)

library(VennDiagram)
set.seed(1) # For reproducibility of results
xx.1 <- list(Bacteria = lst[1][[1]], Plant = lst[2][[1]], 
             Arthropoda=lst[3][[1]]  )
venn.diagram(xx.1, imagetype="tiff",filename ="Venn.tiff", height = 3000, width = 3000,print.mode=("raw"),sigdigs = 1)


```
Extract arthropoda read count:

```{r arthropoda}
#BiocManager::install("edgeR")
#BiocManager::install('limma')
#BiocManager::install('DESeq2')
#BiocManager::install('ctc')
#BiocManager::install('Biobase')

df <- read.table(gzfile(here("merged.readcount.txt.gz")),header=T,stringsAsFactors = F)
dA=df[df$gene_id%in%arth,]
cols=data.frame(names(dA[,-1]))

cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"Bx1B0","NA")
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"Bx1B3",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"Bx2B0",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"Bx2B3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"R32B0",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"R32B3",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"W22B0",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"W22B3",cols$cond)
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"Bx1G0",cols$cond)
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"Bx1G3",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"Bx2G0",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"Bx2G3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"R32G0",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"R32G3",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"W22G0",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"W22G3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G", cols$names.dA....1..) ,"R32G",cols$cond)

cols2=cols[,c(2,1)]
m=c("Bx1G0","Bx1G3","Bx1B0","Bx1B3","W22G0","W22G3","W22B0","W22B3")
combs <- data.frame(t(combn(x = m, m = 2)))
write.table(cols2,"samples_file.txt",row.names = F,col.names = F,sep='\t',quote = F)
write.table(combs,"contrasts.txt",row.names = F,col.names = F,sep='\t',quote = F)
write.table(dA,"count.matrix.arthropoda.txt",row.names = F,sep='\t',quote = F)




system("$TRINITY_HOME/Analysis/DifferentialExpression/run_DE_analysis.pl --matrix count.matrix.arthropoda.txt --method DESeq2 --samples_file samples_file.txt --contrasts contrasts.txt --output Dseq2_components/")

```
