---
title: "Diabroticum Insect Differential expression"
output: github_document
---
The code needs:
1. A classification file for superkingdom	phylum	genus	species	kingdom of organisms (Final_classification.txt)
2. A file specifying annotation of transcripts into superkingdom of origin based on swissprot evidence (annotations_extracted2.txt)
3. A gzipped file of read count per transcript for every sample (merged.readcount.txt.gz)
4. Sample file and contrasts file (samples_file.txt, contrasts.txt)
5. A GO mapping file with every gene mapped to all GO terms - 1:1 mapping (GO_classification_final.dat)
6. Gene length table (Gene_length_from_DEG_annot_xls.txt)



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Classify transcripts into insect, plant and bacteria

```{r classification}

#list.of.packages <- c("here", "VennDiagram","ape","gplots")
#new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
#if(length(new.packages)) install.packages(new.packages, repos = "http://cran.us.r-project.org")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager",repos = "http://cran.us.r-project.org")
library(here)

dat=read.table(here("Final_classification.txt"),header=T,sep='\t',stringsAsFactors = F)

###Load Genus to higher end classification
dat1=unique(dat)
dat1$key=dat1$genus
dat1$key=ifelse(is.na(dat1$key)|dat1$key=="",dat1$species,dat1$key)

dat1$value=dat1$phylum
dat1$value=ifelse(dat1$superkingdom=="Bacteria","Bacteria",dat1$value)
dat1$value=ifelse(dat1$superkingdom=="Archaea","Archaea",dat1$value)
dat1$value=ifelse(dat1$kingdom=="Viridiplantae","Viridiplantae",dat1$value)
dat1$value=ifelse(dat1$kingdom=="Fungi","Fungi",dat1$value)
dat1$value=ifelse(dat1$phylum=="Ascomycota"|dat1$phylum=="Basidiomycota","Fungi",dat1$value)
dat1$value=ifelse(grepl("virus",dat1$key),"Virus",dat1$value)
dat2=dat1[,6:7]
dat3=unique(dat2)
dat4=dat3[dat3$value%in%c("Nematoda","Archaea","Arthropoda","Viridiplantae","Chordata","Fungi"),]
dat4=dat4[dat4$key!="",]

### load modified transcript annotations & process only swissprot evidence
dt=read.table(here("annotations_extracted2.txt"),header=T,sep='\t',stringsAsFactors = F)

dt.x=dt[,c(1,4)]
dt1=dt.x[dt.x$Swissprot.Description!=" NA",]

bact=dt1$Genes[grepl("Bacteria", dt1$Swissprot.Description,ignore.case = TRUE)]
plant=dt1$Genes[grepl("Viridiplantae", dt1$Swissprot.Description,ignore.case = TRUE)]
arthropoda=dt1$Genes[grepl("Arthropoda", dt1$Swissprot.Description,ignore.case = TRUE)]

```

Including Plots Venndiagram


```{r venn, echo=FALSE}
system("rm *.log")
system("rm *.tiff")
library(VennDiagram)
all=unique(dt$Genes)
ann=unique(c(bact,plant,arthropoda))
unk=length(all)-length(ann)
unk
arth=gsub(" ", "", arthropoda)
bac=gsub(" ", "", bact)
pla=gsub(" ", "", plant)

lst=list()
lst[1]=list(bac)
lst[2]=list(pla)
lst[3]=list(arth)

library(VennDiagram)
set.seed(1) # For reproducibility of results
xx.1 <- list(Bacteria = lst[1][[1]], Plant = lst[2][[1]], 
             Arthropoda=lst[3][[1]]  )
venn.diagram(xx.1, imagetype="tiff",filename ="Venn.tiff", height = 3000, width = 3000,print.mode=("raw"),sigdigs = 1)


```

Extract arthropoda read counts per contrast and do differential expression analysis with Deseq2:
```{r arthropoda}
#BiocManager::install("edgeR")
#BiocManager::install('limma')
#BiocManager::install('DESeq2')
#BiocManager::install('ctc')
#BiocManager::install('Biobase')

df <- read.table(gzfile(here("merged.readcount.txt.gz")),header=T,stringsAsFactors = F)
dA=df[df$gene_id%in%arth,]
cols=data.frame(names(dA[,-1]))

cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"Bx1B0","NA")
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"Bx1B3",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"Bx2B0",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"Bx2B3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"R32B0",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"R32B3",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_B0", cols$names.dA....1..) ,"W22B0",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_B3", cols$names.dA....1..) ,"W22B3",cols$cond)
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"Bx1G0",cols$cond)
cols$cond=ifelse(grepl("bx1_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"Bx1G3",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"Bx2G0",cols$cond)
cols$cond=ifelse(grepl("bx2_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"Bx2G3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"R32G0",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"R32G3",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_G0", cols$names.dA....1..) ,"W22G0",cols$cond)
cols$cond=ifelse(grepl("W22_", cols$names.dA....1..) & grepl("_G3", cols$names.dA....1..) ,"W22G3",cols$cond)
cols$cond=ifelse(grepl("R32_", cols$names.dA....1..) & grepl("_G", cols$names.dA....1..) ,"R32G",cols$cond)

cols2=cols[,c(2,1)]
m=c("Bx1G0","Bx1G3","Bx1B0","Bx1B3","W22G0","W22G3","W22B0","W22B3")
combs <- data.frame(t(combn(x = m, m = 2)))
write.table(cols2,"samples_file.txt",row.names = F,col.names = F,sep='\t',quote = F)
#write.table(combs,"contrasts.txt",row.names = F,col.names = F,sep='\t',quote = F)
write.table(dA,"count.matrix.arthropoda.txt",row.names = F,sep='\t',quote = F)

system("$TRINITY_HOME/Analysis/DifferentialExpression/run_DE_analysis.pl --matrix count.matrix.arthropoda.txt --method DESeq2 --samples_file samples_file.txt --contrasts contrasts.txt --output Dseq2_components/")



```

Enrichment analysis for differentially expressed genes:
```{r countDE}
library(goseq)
library(reshape2)
files <- list.files(path="Dseq2_components", pattern="*_results", full.names=TRUE, recursive=FALSE)
#gomapping=read.table("GO_classification1.dat",header=F,sep='\t',fill=T)
#s <- strsplit(as.character(gomapping$V2), ',')
#dat=data.frame(GOterm=rep(gomapping$V1, sapply(s, FUN=length)),director=unlist(s))
#names(dat)=c("GOTerms","Genenames")
#gomap=dat[,c(2,1)]
#write.table(gomap,"GO_classification_final.dat",row.names = F,quote = F,sep='\t')

gomap=read.table(gzfile(here("GO_classification_final.dat.gz")),header=T,sep='\t')
de.genes=gomap$Genenames
gene.len=read.table("Gene_length_from_DEG_annot_xls.txt",header=T,sep='\t',stringsAsFactors = F)

for (x in files)  {
    filenm <- sub(".DESeq2.DE_results", "", x)
    filenm1 <- sub("Dseq2_components/*count.matrix.", "", filenm)
    print(paste("PROCESSING GENE ENRICHMENT FOR:",filenm1,sep=' '))
    cat("--------------------------------------\n")
    t <- read.table(x, header=TRUE, stringsAsFactors = F) # load file
    #filter based on padj<0.05 and two-fold change
    t1 <- t[t$padj<0.05&abs(t$log2FoldChange)>1,]
    # apply function
    assayed.genes <- rownames(t)
    de.genes <- rownames(t1)
    gene.vector=as.integer(assayed.genes%in%de.genes)
    names(gene.vector)=assayed.genes
    gene_len=gene.len[gene.len$gene_id%in%assayed.genes,]
    gene.length<-as.vector(t(gene_len$Gene.Length))
    names(gene.length)=gene_len$gene_id
    
    pwf<-nullp(gene.vector,bias.data=gene.length)
    GO.wall=goseq(pwf,gene2cat=gomap)
    GO.samp=goseq(pwf,gene2cat=gomap,method="Sampling",repcnt=1000)
     plot(log10(GO.wall[,2]),log10(GO.samp[match(GO.samp[,1],GO.wall[,1]),2]),xlab=paste(filenm1,"log10(Wallenius p-values)",sep=' '), ylab="log10(Sampling p-values)", xlim=c(-3,0) )
    abline(0,1,col=3,lty=2)
    
    enriched.GO=GO.wall$category[p.adjust(GO.wall$over_represented_pvalue, method="BH")<.05]
enriched.GO.tab=GO.wall[p.adjust(GO.wall$over_represented_pvalue, method="BH")<.05,]
library(GO.db)
GOtable=paste(filenm1,"enriched_GOtables.txt")
dir.create("overallGOtables")
write.table(enriched.GO.tab,paste("overallGOtables/",GOtable,sep=''),row.names = F,quote = F,sep='\t')
    

 
#Plotting the pvalues against one another, we see that there is little difference between the two methods. However, the accuracy of the sampling method is limited by the number of samples generated, repcnt, such that very low p-values will not be correctly calculated. Significantly enriched GO terms may then be missed after correcting for multiple testing.

#Testing without length correction. NEVER use this option as your final analysis.

GO.nobias=goseq(pwf,gene2cat=gomap,method="Hypergeometric")
  

    {plot(log10(GO.wall[,2]),log10(GO.nobias[match(GO.nobias[,1],GO.wall[,1]),2]),xlab="log10(Wallenius p-values)", ylab="log10(Nobias p-values)", xlim=c(-3,0) )
    abline(0,1,col=3,lty=2)}

    
#Ignoring length bias gives very different results from a length bias corrected analysis.

#Enriched Go categories:


GOdetails=paste(filenm1,"enriched_GOdetails.txt")
dir.create("overallGOdetails")
sink(file=paste("overallGOdetails/",GOdetails,sep=''),type="output")
cat("AllGO: ",enriched.GO)
for(go in enriched.GO[1:length(enriched.GO)]){
  print(GOTERM[[go]])
  cat("--------------------------------------\n")
}

sink()


#Kegg Pathway:
#keggmapping=read.table("Kegg_pathway.dat",header=T,sep='\t',fill=T)
#s <- strsplit(as.character(keggmapping$Gene.IDs), ',')
#dat=data.frame(Keggterm=rep(keggmapping$Pathway.ID,sapply(s,FUN=length)),director=unlist(s))
#names(dat)=c("KeggTerms","Genenames")
#keggmap=dat[,c(2,1)]
#write.table(keggmap,"KEGG_classification_final.txt",row.names = F,quote = F,sep='\t')
keggmap=read.table(here("KEGG_classification_final.txt"),header=T,sep='\t')

KEGG=goseq(pwf,gene2cat=keggmap)
enriched.kegg=KEGG$category[p.adjust(KEGG$over_represented_pvalue, method="BH")<.05]
enriched.kegg.tab=KEGG[p.adjust(KEGG$over_represented_pvalue, method="BH")<.05,]
keggtable=paste(filenm1,"Keggtables.txt")
dir.create("overallKeggtables")
write.table(KEGG,paste("overallkeggtables/",keggtable,sep=''),row.names = F,quote = F,sep='\t')

 
}

```